wpilibcExamples/src/main/cpp/examples/StateSpaceDifferentialDriveSimulation/cpp/subsystems/DriveSubsystem.cpp

```diff
@@ -65,10 +69,46 @@ void DriveSubsystem::ArcadeDrive(double fwd, double rot) {
   m_drive.ArcadeDrive(fwd, rot);
 }
 
-void DriveSubsystem::TankDriveVolts(units::volt_t left, units::volt_t right) {
-  m_leftMotors.SetVoltage(left);
-  m_rightMotors.SetVoltage(-right);
+void DriveSubsystem::FollowState(const frc::Trajectory::State& targetState) {
+  // Calculate via Ramsete the speed vector needed to reach the target state
+  frc::ChassisSpeeds speedVector =
+      m_ramseteController.Calculate(m_odometry.GetPose(), targetState);
+  // Convert the vector to the separate velocities of each side of the
+  // drivetrain
+  frc::DifferentialDriveWheelSpeeds wheelSpeeds =
+      kDriveKinematics.ToWheelSpeeds(speedVector);
+
+  // PID and Feedforward control
+  // This can be replaced by calls to smart motor controller closed-loop control
+  // functionality. For example:
+  // mySmartMotorController.setSetpoint(leftMetersPerSecond,
+  // ControlType.Velocity);
+
+  auto leftFeedforward = m_feedforward.Calculate(
+      wheelSpeeds.left, (wheelSpeeds.left - m_previousSpeeds.left) /
+                            (targetState.t - m_previousTime));
```

Comment:
There are two problems with this approach:
 - `targetState.t - m_previousTime` will be -1 on the first call to this method, potentially resulting in negative acceleration when positive is what you actually want.
 - `targetState.t - m_previousTime` will be a large value when you start to follow another trajectory because `m_previousTime` will contain the duration of the previously run trajectory.

Reply:
For reference, LinearPlantInversionFeedforward takes a nominal dt in the ctor, then has Calculate() just take the current velocity and next velocity. The input change caused by dt jitter has been negligible in my experience.

SimpleMotorFeedforward::Calculate() takes a dt because the ctor didn't. I'd recommend just throwing in a nominal dt here to massively simplify things.

Reply:
Is there a reason that `SimpleMotorFeedforward` can't be altered to take a `dt` in the ctor?

Assuming I use the overload that you're suggesting, how would I get the next velocity? Or is it the target velocity (if so, I suggest renaming the parameter- it's confusing)?

Reply:
> Is there a reason that SimpleMotorFeedforward can't be altered to take a dt in the ctor?

It could if we're OK with breaking changes in the API.

> Assuming I use the overload that you're suggesting, how would I get the next velocity? Or is it the target velocity (if so, I suggest renaming the parameter- it's confusing)?

To compute a feedforward, you use where you wanted to be at the current timestep, where you want to be at the next timestep, and information about your dynamics. In this case, you can use previous and current target velocity. You'll lag behind by a timestep, but that's OK as long as the controller thinks the previous target velocity is where it's trying to get to.

Reply:
Is it a problem using `(previous, current)` for FF and `(current)` for PID?

Reply:
Doing so is technically incorrect, but you wouldn't notice much of a difference unless you're looking for perfect tracking. If it's easy to just use previous, I'd do that as long as we can ensure AtGoal() and other things are returning the right thing.